### PR TITLE
[search] Index "strasse" both together and separately with street name

### PR DESCRIPTION
**Pull request by tatiana-yan**

### CODE REVIEW

File: base/base_tests/string_utils_test.cpp [hunk 1]
```diff
@@ -549,16 +549,29 @@ UNIT_TEST(StartsWith)
...
+  }
+  {
+    auto const s = MakeUniString("xyz");
+    TEST(EndsWith(s, MakeUniString("")), ());
+    TEST(EndsWith(s, MakeUniString("z")), ());
+    TEST(EndsWith(s, MakeUniString("yz")), ());
+    TEST(EndsWith(s, MakeUniString("xyz")), ());
+    TEST(!EndsWith(s, MakeUniString("abcxyz")), ());
+    TEST(!EndsWith(s, MakeUniString("ayz")), ());
+    TEST(!EndsWith(s, MakeUniString("axyz")), ());
```
Comment (mpimenov) [inline base/base_tests/string_utils_test.cpp]: Test non-ascii too, please.

File: generator/search_index_builder.cpp [hunk 2]
```diff
@@ -184,8 +184,26 @@ struct FeatureNameInserter
                                         {
                                           AddToken(lang, token);
                                         });
-      for (auto const & token : tokens)
+      for (size_t i = 0; i < tokens.size(); ++i)
+      {
+        auto const & token = tokens[i];
         filter.Put(token, false /* isPrefix */, 0 /* tag */);
+
+        auto static const kStrasse = strings::MakeUniString("strasse");
```
Comment (mpimenov) [inline generator/search_index_builder.cpp]: Can we factor this out into a function?

**Comment by tatiana-yan** [comment]
JTMAC

File: generator/search_index_builder.cpp (hunk 2 above)
Comment (tatiana-yan) [reply generator/search_index_builder.cpp]: done

File: base/base_tests/string_utils_test.cpp (hunk 1 above)
Comment (tatiana-yan) [reply base/base_tests/string_utils_test.cpp]: done

File: generator/search_index_builder.cpp [hunk 3]
```diff
@@ -153,6 +153,29 @@ struct FeatureNameInserter
     m_keyValuePairs.emplace_back(key, m_val);
   }
 
+  void AddStrasseNames(search::QueryTokens const & tokens, signed char lang) const
```
Comment (mpimenov) [inline generator/search_index_builder.cpp]: (lang, tokens)
as in other methods

File: base/string_utils.cpp [hunk 4]
```diff
@@ -280,6 +280,20 @@ bool StartsWith(std::string const & s1, char const * s2) { return (s1.compare(0,
 
 bool StartsWith(std::string const & s1, std::string const & s2) { return (s1.compare(0, s2.length(), s2) == 0); }
 
+bool EndsWith(UniString const & s1, UniString const & s2)
+{
+  if (s1.size() < s2.size())
+    return false;
+
+  size_t const start = s1.size() - s2.size();
```
Comment (mpimenov) [inline base/string_utils.cpp]: nit: std::equal would be more in line with the implementations of other overrides

Comment (tatiana-yan) [reply base/string_utils.cpp]: done

File: generator/search_index_builder.cpp (hunk 3 above)
Comment (tatiana-yan) [reply generator/search_index_builder.cpp]: done

File: base/string_utils.cpp [hunk 5]
```diff
@@ -280,6 +280,15 @@ bool StartsWith(std::string const & s1, char const * s2) { return (s1.compare(0,
...
+  return std::equal(s1.begin() + start, s1.end(), s2.begin());
```
Comment (mpimenov) [inline base/string_utils.cpp]: s1.end() - s2.size()?

(and no start variable)

Comment (tatiana-yan) [reply base/string_utils.cpp]: done

**Pull request merged**